File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int minimum=nums[0],maximum=nums[0],ans=nums[0];
        for(int i=1;i<nums.size();i++){
            if(nums[i]<0){
               swap(maximum,minimum);
            }
            
            maximum = max(nums[i],maximum*nums[i]);
            minimum = min(nums[i],minimum*nums[i]);
            ans =  max(ans,maximum);
        }
        
        return ans;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        ans,max_streak,min_streak = nums[0],nums[0],nums[0]
        
        for i in range(1,n):
            cur = nums[i]
            new_max_streak = max(cur,cur*max_streak,cur*min_streak)
            min_streak = min(cur,cur*max_streak,cur*min_streak)
            max_streak = new_max_streak
            ans = max(ans,max_streak)
        return ans
            